src/engraving/libmscore/navigate.cpp

```diff
@@ -760,8 +818,61 @@ EngravingItem* Score::prevElement()
         case ElementType::BAR_LINE: {
             for (; e && e->type() != ElementType::SEGMENT; e = e->parentItem()) {
             }
-            Segment* s = toSegment(e);
-            return s->prevElement(staffId);
+            EngravingItem* previousElement = toSegment(e)->prevElement(staffId);
+
+            if (previousElement->type() != ElementType::VBOX
+                && previousElement->type() != ElementType::HBOX
+                && previousElement->type() != ElementType::TBOX) {
+                return previousElement;
+            }
+
+            EngravingObjectList list = previousElement->children();
+
+            auto compareFunction = [](const std::pair<EngravingItem*, int>& firstChild,
+                                      const std::pair<EngravingItem*, int>& secondChild) {
+                    PointF position1 = firstChild.first->pos() + firstChild.first->bbox().center();
+                    PointF position2 = secondChild.first->pos() + secondChild.first->bbox().center();
+
+                    if (position2.y() - position1.y() > 10) { // Also automatically test if position2.y() > position1.y()
+                        return true;
+                    } else if (position1.y() - position2.y() > 10) { // Also automatically test if position1.y() > position2.y()
+                        return false;
+                    }
+
+                    if (position1.x() < position2.x()) {
+                        return true;
+                    } else if (position1.x() > position2.x()) {
+                        return false;
+                    }
+
+                    if (firstChild.second < secondChild.second) {
+                        return true;
+                    } else {
+                        return false;
+                    }
+                };
+
+            std::set<std::pair<EngravingItem*, int>, decltype(compareFunction)> boxChildren(compareFunction);
+            int index = 0;
+
+            for (EngravingObject* object : list) {
+                EngravingItem* engravingItem = toEngravingItem(object);
+
+                if (!engravingItem) {
+                    continue;
+                }
+
+                boxChildren.insert(std::pair<EngravingItem*, int>(engravingItem, index));
+                index++;
+            }
```

Comment:
Most of this code is identical to `nextElement()` so it should be factored out into a separate `static` function that takes the box as an argument and returns the set of child+index pairs. You can use the following declarations to make the code neater:

```c++
using ElementPair = std::pair<EngravingItem*, int>; // element and its index in parent

static bool compareElementPairs(const ElementPair& child1, const ElementPair& child2)
{
    PointF p1 = ...;
    PointF p2 = ...;
    // etc.
}

static std::set<ElementPair, decltype(&compareFunction)> childPairsSet(const EngravingItem* element)
{
    std::set<ElementPair, decltype(&compareFunction)> children(&compareElementPairs);

   // add elements to set

   return children;
}
```

Put that somewhere near the top of `navigate.cpp` outside of any functions, then inside `nextElement()` and `previousElement()` you just have to do:

```c++
auto boxChildren = childPairsSet(e);
// now use the set
```